include/sycl_ext_complex.hpp

```diff
@@ -1232,6 +1240,436 @@ _SYCL_EXT_CPLX_INLINE_VISIBILITY complex<_Tp> tan(const complex<_Tp> &__x) {
+      rtn[i] = sycl::ext::cplx::math_func(x[i]);                               \
+                                                                               \
+    return rtn;                                                                \
+  }
+
+MATH_OP_ONE_PARAM(abs, T, complex<T>);
+MATH_OP_ONE_PARAM(acos, complex<T>, complex<T>);
+MATH_OP_ONE_PARAM(asin, complex<T>, complex<T>);
+MATH_OP_ONE_PARAM(atan, complex<T>, complex<T>);
+MATH_OP_ONE_PARAM(acosh, complex<T>, complex<T>);
+MATH_OP_ONE_PARAM(asinh, complex<T>, complex<T>);
+MATH_OP_ONE_PARAM(atanh, complex<T>, complex<T>);
+MATH_OP_ONE_PARAM(arg, T, complex<T>);
+MATH_OP_ONE_PARAM(conj, complex<T>, complex<T>);
+MATH_OP_ONE_PARAM(cos, complex<T>, complex<T>);
+MATH_OP_ONE_PARAM(cosh, complex<T>, complex<T>);
+MATH_OP_ONE_PARAM(exp, complex<T>, complex<T>);
+MATH_OP_ONE_PARAM(log, complex<T>, complex<T>);
+MATH_OP_ONE_PARAM(log10, complex<T>, complex<T>);
+MATH_OP_ONE_PARAM(norm, T, complex<T>);
+MATH_OP_ONE_PARAM(proj, complex<T>, complex<T>);
+MATH_OP_ONE_PARAM(proj, complex<T>, T);
+MATH_OP_ONE_PARAM(sin, complex<T>, complex<T>);
+MATH_OP_ONE_PARAM(sinh, complex<T>, complex<T>);
+MATH_OP_ONE_PARAM(sqrt, complex<T>, complex<T>);
+MATH_OP_ONE_PARAM(tan, complex<T>, complex<T>);
+MATH_OP_ONE_PARAM(tanh, complex<T>, complex<T>);
+
+#undef MATH_OP_ONE_PARAM
+
+#define MATH_OP_TWO_PARAM(math_func, rtn_type, arg_type1, arg_type2)           \
+  template <typename T, std::size_t NumElements,                               \
+            typename = std::enable_if<is_genfloat<T>::value ||                 \
+                                      is_gencomplex<T>::value>>                \
+  _SYCL_EXT_CPLX_INLINE_VISIBILITY sycl::marray<rtn_type, NumElements>         \
+  math_func(const sycl::marray<arg_type1, NumElements> &x,                     \
+            const sycl::marray<arg_type2, NumElements> &y) {                   \
+    sycl::marray<rtn_type, NumElements> rtn;                                   \
+    for (std::size_t i = 0; i < NumElements; ++i)                              \
+      rtn[i] = sycl::ext::cplx::math_func(x[i], y[i]);                         \
+                                                                               \
+    return rtn;                                                                \
+  }                                                                            \
+                                                                               \
+  template <typename T, std::size_t NumElements,                               \
+            typename = std::enable_if<is_genfloat<T>::value ||                 \
+                                      is_gencomplex<T>::value>>                \
+  _SYCL_EXT_CPLX_INLINE_VISIBILITY sycl::marray<rtn_type, NumElements>         \
+  math_func(const sycl::marray<arg_type1, NumElements> &x,                     \
+            const arg_type2 &y) {                                              \
```

Comment:
just to confirm, we allow `pow(marray<T>{a}, T b)` ? I have no idea if it's a nice feature or not...

Reply:
Following our conversation, I did not think you meant: `marray<T>, T` or `T, marray<T>`.

No, there is not an overload on marray which invokes the "regular" pow function.

Here are the pow overloaded for marray:

```
marray<complex<T>> pow(const marray<complex<T>> &x, const marray<complex<T>> &y);

marray<complex<T>> pow(const marray<complex<T>> &x, const complex<T> &y);
marray<complex<T>> pow(const complex<T> &x, const marray<complex<T>> &y);

marray<complex<T>> pow(const marray<complex<T>> &x, const marray<T> &y);
marray<complex<T>> pow(const marray<T> &x, const marray<complex<T>> &y);

marray<complex<T>> pow(const complex<T> &x, const marray<T> &y);
marray<complex<T>> pow(const marray<T> &x, const complex<T> &y);

marray<complex<T>> pow(const marray<complex<T>> &x, T y);
marray<complex<T>> pow(T x, const marray<complex<T>> &y);
```

Do you still want to trim the pow interface?

Reply:
Sorry for the delay; I should improve my bookkeeping strategy...

I think I want to keep this one to start with:
>marray<complex<T>> pow(const marray<complex<T>> &x, const marray<complex<T>> &y);